localstack/services/apigateway/provider_asf.py

```diff
@@ -0,0 +1,185 @@
+        """
+        Removes the given rest api.
+        :param rest_api_id: the rest api to remove
+        """
+        rules = self.router_rules.pop(rest_api_id, [])
+        for rule in rules:
+            self.router.remove_rule(rule)
+
+    def _restapis_handler(self, request: Request, path=None) -> Response:
+        invocation_context = to_invocation_context(request)
+
+        result = invoke_rest_api_from_request(invocation_context)
+        if result is not None:
+            if isinstance(result, LambdaResponse):
+                headers = Headers(dict(result.headers))
+                for k, values in result.multi_value_headers.items():
+                    for value in values:
+                        headers.add(k, value)
+            else:
+                headers = dict(result.headers)
+            return Response(
+                response=result.content,
+                status=result.status_code,
+                headers=headers,
+            )
+
+        raise NotFound()
+
+    def _restapis_user_request_handler(self, request: Request, stage=None, path=None):
+        return self._restapis_handler(request, path)
+
+    def _restapis_host_handler(self, request: Request, path=None, server=None) -> Response:
+        return self._restapis_handler(request)
+
+
+class AsfApigatewayProvider(ApigatewayProvider):
+    """Modern ASF provider that uses an ApigatewayRouter to dispatch requests to user routes."""
+
+    router: ApigatewayRouter
+
+    def __init__(self, router: ApigatewayRouter = None):
+        self.router = router or ApigatewayRouter(router=ROUTER)
+
+    def create_rest_api(self, context: RequestContext, request: CreateRestApiRequest) -> RestApi:
+        result: RestApi = super().create_rest_api(context, request)
+        API_REGIONS[result["id"]] = context.region
+        self.router.add_rest_api(result)
+        return result
+
+    def delete_rest_api(self, context: RequestContext, rest_api_id: String) -> None:
+        super().delete_rest_api(context, rest_api_id)
+        self.router.remove_rest_api(rest_api_id)
+
+    @handler("TestInvokeMethod", expand=False)
+    def test_invoke_method(
+        self, context: RequestContext, request: TestInvokeMethodRequest
+    ) -> TestInvokeMethodResponse:
+
+        invocation_context = to_invocation_context(context.request)
+        invocation_context.method = request["httpMethod"]
+
+        if data := parse_json_or_yaml(to_str(invocation_context.data or b"")):
+            orig_data = data
+            path_with_query_string = orig_data.get("pathWithQueryString", None)
+            if path_with_query_string:
+                invocation_context.path_with_query_string = path_with_query_string
+            invocation_context.data = data.get("body")
+            invocation_context.headers = orig_data.get("headers", {})
+
+        result = invoke_rest_api_from_request(invocation_context)
+
+        # FIXME: there are also multi-value-headers, log, and latency
```

Comment:
Could you shortly elaborate on this `FIXME` comment?

Reply:
extended the explanation